src/log.rs

```diff
@@ -0,0 +1,128 @@
+//! This module implements event builders and streams for retrieving events
+//! emitted by a contract.
+
+use crate::errors::ExecutionError;
+use futures::compat::{Compat01As03, Future01CompatExt, Stream01CompatExt};
+use futures::ready;
+use futures::stream::Stream;
+use pin_project::{pin_project, project};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use web3::api::{CreateFilter, FilterStream, Web3};
+use web3::types::{Filter, Log};
+use web3::Transport;
+
+/// A log stream that emits logs matching a certain filter.
+///
+/// Note that when creating a log stream that is only valid until a certain
+/// block number, the `Stream` implementation will currently remain in the
+/// pending state indefinitely.
+#[pin_project]
+pub struct LogStream<T: Transport> {
+    #[pin]
+    state: LogStreamState<T>,
+}
+
+/// The state of the log stream. It can either be creating a new log filter for
+/// retrieving new logs or streaming logs from the created filter.
+#[pin_project]
+enum LogStreamState<T: Transport> {
+    CreatingFilter(#[pin] CompatCreateFilter<T, Log>, Duration),
+    Streaming(#[pin] CompatFilterStream<T, Log>),
+}
+
+impl<T: Transport> LogStream<T> {
+    /// Create a new log stream from a given web3 provider, filter and polling
+    /// parameters.
+    pub fn new(web3: Web3<T>, filter: Filter, poll_interval: Duration) -> Self {
+        let create_filter = web3.eth_filter().create_logs_filter(filter).compat();
+        let state = LogStreamState::CreatingFilter(create_filter, poll_interval);
+        LogStream { state }
+    }
+}
+
+impl<T: Transport> Stream for LogStream<T> {
+    type Item = Result<Log, ExecutionError>;
+
+    #[project]
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            let mut state = self.as_mut().project().state;
+
+            #[project]
+            let next_state = match state.as_mut().project() {
+                LogStreamState::CreatingFilter(create_filter, poll_interval) => {
+                    let log_filter = match ready!(create_filter.poll(cx)) {
+                        Ok(log_filter) => log_filter,
+                        Err(err) => return Poll::Ready(Some(Err(err.into()))),
+                    };
+                    let stream = log_filter.stream(*poll_interval).compat();
+                    LogStreamState::Streaming(stream)
+                }
+                LogStreamState::Streaming(stream) => {
+                    return stream
+                        .poll_next(cx)
+                        .map(|result| result.map(|log| Ok(log?)))
```

Comment:
Took me a while to understand that `Ok(log?)` is syntactic sugar for `log.map_err(ExecutionError::from)`

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2FxT0xeJpnrWC4XWblEk%2Fgiphy-downsized-medium.gif&token=mind_blown" data-gitmeme-token="mind_blown"><img src="https://media2.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy-downsized-medium.gif" title="Created by gitme.me with /mind_blown"/></a>